flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/JobLeaderService.java

```diff
@@ -251,90 +254,83 @@ private JobManagerLeaderListener(JobID jobId) {
 		}
 
 		public void stop() {
-			stopped = true;
+			synchronized (lock) {
+				stopped = true;
 
-			if (rpcConnection != null) {
-				rpcConnection.close();
+				if (rpcConnection != null) {
+					rpcConnection.close();
+				}
 			}
 		}
 
 		public void reconnect() {
-			if (stopped) {
-				LOG.debug("Cannot reconnect because the JobManagerLeaderListener has already been stopped.");
-			} else {
-				final RegisteredRpcConnection<JobMasterId, JobMasterGateway, JMTMRegistrationSuccess> currentRpcConnection = rpcConnection;
-
-				if (currentRpcConnection != null) {
-					if (currentRpcConnection.isConnected()) {
-
-						if (currentRpcConnection.tryReconnect()) {
-							// double check for concurrent stop operation
-							if (stopped) {
-								currentRpcConnection.close();
-							}
-						} else {
-							LOG.debug("Could not reconnect to the JobMaster {}.", currentRpcConnection.getTargetAddress());
-						}
+			synchronized (lock) {
+				if (stopped) {
+					LOG.debug("Cannot reconnect because the JobManagerLeaderListener has already been stopped.");
+				} else {
+					if (rpcConnection != null) {
+						Preconditions.checkState(
+							rpcConnection.tryReconnect(),
```

Comment:
Can we add a javadoc to `RegisteredRpcConnection#tryReconnect` stating that it only returns false in the case of a concurrent modification?

Reply:
Yes, good idea. I'll add it.